go/vt/vttablet/tabletmanager/vreplication/vcopier.go

```diff
@@ -264,86 +443,183 @@ func (vc *vcopier) copyTable(ctx context.Context, tableName string, copyState ma
 				TableName: initialPlan.SendRule.Match,
 			}
 			fieldEvent.Fields = append(fieldEvent.Fields, rows.Fields...)
-			vc.tablePlan, err = plan.buildExecutionPlan(fieldEvent)
+			tablePlan, err := plan.buildExecutionPlan(fieldEvent)
 			if err != nil {
 				return err
 			}
 			pkfields = append(pkfields, rows.Pkfields...)
 			buf := sqlparser.NewTrackedBuffer(nil)
-			buf.Myprintf("update _vt.copy_state set lastpk=%a where vrepl_id=%s and table_name=%s", ":lastpk", strconv.Itoa(int(vc.vr.id)), encodeString(tableName))
-			updateCopyState = buf.ParsedQuery()
+			buf.Myprintf(
+				"update _vt.copy_state set lastpk=%a where vrepl_id=%s and table_name=%s", ":lastpk",
+				strconv.Itoa(int(vc.vr.id)),
+				encodeString(tableName),
+			)
+			copyStateUpdate := buf.ParsedQuery()
+			copyWorkQueue.open(copyStateUpdate, pkfields, tablePlan)
 		}
 		if len(rows.Rows) == 0 {
 			return nil
 		}
 
-		// The number of rows we receive depends on the packet size set
-		// for the row streamer. Since the packet size is roughly equivalent
-		// to data size, this should map to a uniform amount of pages affected
-		// per statement. A packet size of 30K will roughly translate to 8
-		// mysql pages of 4K each.
-		if err := vc.vr.dbClient.Begin(); err != nil {
-			return err
+		// Clone rows, since pointer values will change while async work is
```

Comment:
@maxenglander, should we optimize this by not cloning, in case we have only one worker there is no parallelism?

Reply:
Yes will do